### DISCUSSION TITLE
Support for HTML interactive or inline output in .qmd file

### DISCUSSION THREAD

**Discussion by anbrav0** [posit-dev/positron#4033] [Feedback] [answered]
Currently in Positron, when skimming through data, the output is shown in the console like so: 
![image](https://github.com/user-attachments/assets/5dbdd8fc-c9ca-4861-8773-03271bc6e3b6)

in RStudios, you have the option to view the data interactively, through inline like so:
![image](https://github.com/user-attachments/assets/0b1c446d-642b-4c2d-93c0-12d0b39ead67)

I see an issue for future support HTML outputs in notebooks has been opened at #4018
and I understand there is continued talk about this function here #1533 and here #1418 and I think also relevant here #3718

I am wondering if similar inline output can be added to .qmd files? 

Thank you to the team for all your work on this project! 

Positron version
```
Positron Version: 2024.07.0 (system setup) build 2024.07.0-21
Code - OSS Version: 1.90.0
Commit: 82b335e27545e804b7fd806db00477a6eb49951d
Date: 2024-07-09T03:17:59.552Z
Electron: 29.4.0
Chromium: 122.0.6261.156
Node.js: 20.9.0
V8: 12.2.281.27-electron.0
OS: Windows_NT x64 10.0.19045

```

**Comment by juliasilge** [comment]
We don't have immediate plans for inline HTML output for Quarto, because that is the same as supporting general inline support. As of today, our recommendation is:

- If you want an inline experience, use `.ipynb`
- If you want a send-to-console experience, use `.qmd`

Like you found in those various discussions, people definitely have strong feelings about wanting the "other" option (whatever "other" means for them in particular) and it's really helpful for us to know what is most needed/preferred. During just our first few weeks of having Positron available, this is coming up as a really top priority from many users, which is really interesting and helpful for us in planning. We appreciate you and others sharing your preferences! 🙌

**Reply by ntluong95** [reply]
A lot of ppl request for in-line output of .qmd file in Positron I think. I see very few against it

**Reply by juliasilge** [reply]
Because those people have exactly what they want right now! 😄

**Reply by ntluong95** [reply]
I dont think so :( Folks have been working with Rstudio all remember the in-line output

**Reply by kldhana** [reply]
@juliasilge, are there any plans to change the viewer panel (i.e., R HTML widget) on the right panel to match the theme's color?

**Reply by stefanocoretta** [reply]
I don't see how anybody who does any kind of proper data analysis for academic research handles complex statistical analysis without inline output. 🤣 Of course I recognise that some might just need to make a couple plots and a bunch of t-tests per project.

For my research and teaching of stats in linguistics, having hundreds of plots in the Plot panel is unthinkable. (I care less about other types of output although I rely on that too).

Plots are already captured by the plot panel so it's not a wild idea that they be captured by the editor.

**Comment by stefanocoretta** [comment]
My deal breaker for switching to Positron is inline output support (whether it's code output, data, plots, anything else). My workflow really depends on it, so I wouldn't be able to get the most from Positron (although I would love to switch because the RStudio editor _still_ does not correctly display certain characters, https://github.com/rstudio/rstudio/issues/14900)

**Reply by stefanocoretta** [reply]
To expand on this (after reading other issues), I do a lot of iterative data analysis and I base my workflow on Quarto files where I read, check, mutate, plot, model data. Having the output of code "live" inline is fundamental because it allows me to see the context of the output. Moreover, I can jump from one code chunk to the other and see its output.

Currently in Positron when using a Quarto file you have either to rerun the code or for plots remember which plot corresponds to which code (or rerun the code for the plot, which can be time consuming for certain complex plots). This is impractical in a normal situation of iterative analysis, where sometimes I can have tens of plots in a single document. Jupyter notebooks include the output in the document file itself and they are not "plain text", making them less ideal for versioning.

I've been teaching Rmd/Quarto to students for the past four years both at UG and PG level, and one thing they said they loved was indeed the inline output (even those students who were used to python and Jupyter notebooks said they liked the UX of Quarto files). Of course, not everyone likes live inline output, but that doesn't mean it should not be implemented (why yes for jupyter notebooks, but no for Quarto?).

Whether the Quarto file is linked to it's own console, the general console, no console at all, whether the code goes to the console and the output is captured, or otherwise, I personally (and I think most of us who are used to do their research with this workflow in Quarto) do not care. I also don't care if this has to be "activated" per document, per project, or generally. I just believe it's not a wise move to offer one great feature for years and then take it away from us. :D

**Reply by AlbertoFabbri93** [reply]
I completely agree, and my workflow is essentially the same as @stefanocoretta’s. For this type of iterative analysis—where it's necessary to view multiple plots, switch back and forth between them, and make adjustments—it’s crucial to have inline output, including support for interactive HTML plots.

I also share the concern about version control in Jupyter, which is why I try to avoid using it when possible. I truly hope you’ll consider enabling inline output in Positron, similar to what we currently have in RStudio. This feature would significantly enhance the analysis experience.

**Reply by Santoshjnawali** [reply]
I have the same thing to share! In-line output truly enhances the analysis experience with Quarto. It streamlines the process and makes the workflow more intuitive and interactive. It also helps in developing a prototype for a report without rendering every time.

**Comment by VLucet** [comment]
The last exchange on this is almost a year old, has there been any update toward _wanting_ to implement this feature to bring Positron on par with RStudio? For me who teaches R, it is the last roadblock toward switching my students (who are used to inline rmarkdown or quarto in RStudio) over to the more rigorous Positron.

**Reply by jmcphers** [reply, marked as answer]
Yes, we have basic support for this now! It's currently a preview feature, so you'll need to opt in here:

<img width="438" height="124" alt="image" src="https://github.com/user-attachments/assets/4b35b1e5-f430-4e6d-a4bb-7fd2e7f039c3" />

See https://github.com/posit-dev/positron/pull/12036 for a longer description of what's currently implemented and planned.

**Reply by juliasilge** [reply]
I will chime that we don't currently expect this feature to be polished enough today for teaching or learners. If inline output is important for your teaching plans, I would recommend sticking with RStudio until probably later this year. It will likely be ready for students by the fall semester.

**Reply by AlbertoFabbri93** [reply]
@jmcphers I do not see this option in Positron 2026.04.1. Is it only available in the beta version, behind a flag, or?

**Reply by juliasilge** [reply]
It's there! The name of the setting is `positron.quarto.inlineOutput.enabled`:

<img width="727" height="518" alt="Screenshot 2026-04-21 at 8 08 22 AM" src="https://github.com/user-attachments/assets/84060a0b-80b7-43ee-9510-cdb0ed7a1dd6" />
